lessons/lesson5.md

```diff
@@ -0,0 +1,234 @@
+
+```yaml
+send-promo:
+  function: actions/send-promo/index.js
+  web: 'yes'
+  runtime: 'nodejs:10'
+  inputs:
+    LOG_LEVEL: debug
+    tenant: $CAMPAIGN_STANDARD_TENANT
+    apiKey: $CAMPAIGN_STANDARD_API_KEY
+    workflowId: $CAMPAIGN_STANDARD_WORKFLOW_ID
+  annotations:
+    require-adobe-auth: true
+    final: true
+```
+
+Then you should update the source code at `actions/send-promo/index.js` as following:
+
+```javascript
+/**
+ * This action triggers Campaign Standard workflow to send promotion email to a specific email address
+ */
+
+const { Core } = require('@adobe/aio-sdk')
+const { CampaignStandard } = require('@adobe/aio-sdk')
+const { errorResponse, getBearerToken, stringParameters, checkMissingRequestInputs } = require('../utils')
+
+// main function that will be executed by Adobe I/O Runtime
+async function main (params) {
+  // create a Logger
+  const logger = Core.Logger('main', { level: params.LOG_LEVEL || 'info' })
+
+  try {
+    // 'info' is the default level if not set
+    logger.info('Calling the main action')
+
+    // log parameters, only if params.LOG_LEVEL === 'debug'
+    logger.debug(stringParameters(params))
+
+    // check for missing request input parameters and headers
+    const requiredParams = ['apiKey', 'tenant', 'workflowId', 'email']
+    const errorMessage = checkMissingRequestInputs(params, requiredParams, ['Authorization'])
+    if (errorMessage) {
+      // return and log client errors
+      return errorResponse(400, errorMessage, logger)
+    }
+
+    // extract the user Bearer token from the input request parameters
+    const token = getBearerToken(params)
+
+    // initialize the sdk
+    const campaignClient = await CampaignStandard.init(params.tenant, params.apiKey, token)
+
+    // get workflow from Campaign Standard
+    const workflow = await campaignClient.getWorkflow(params.workflowId)
+    const wkfHref = workflow.body.activities.activity.signal1.trigger.href
+
+    // trigger the signal activity API
+    const triggerResult = await campaignClient.triggerSignalActivity(wkfHref, { source: 'API', parameters: { email: params.email } })
+    
+    // log the trigger result
+    logger.info(triggerResult)
+
+    const response = {
+      statusCode: 200,
+      body: { success: 'ok' }
+    }
+
+    // log the response status code
+    logger.info(`${response.statusCode}: successful request`)
+    return response
+  } catch (error) {
+    // log any server errors
+    logger.error(error)
+    // return with 500
+    return errorResponse(500, 'server error', logger)
+  }
+}
+
+exports.main = main
+```
+
+To update the UI, open `App.js` and add method `sendPromo()` as following. Also, don't forget to bind this method inside the constructor: `this.sendPromo = this.sendPromo.bind(this)`.
+
+```javascript
+async sendPromo (email) {
+  try {
+    if (window.confirm(`Send promo code to ${email}?`)) {
```

Comment:
Maybe you can use https://react-spectrum.adobe.com/react-spectrum/AlertDialog.html here instead ?

Reply:
good point, I was not aware of it, will try out and update if it works